File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/protocol_sdk/IDOMeasureManager.swift
//
//  IDOMeasureManager.swift
//  protocol_channel
//
//  Created by hc on 2026/3/19.
//

import Foundation

private var _measureMgr: Measure? {
    return SwiftProtocolChannelPlugin.shared.measure
}

private var _measureImpl: MeasureDelegateImpl {
    return MeasureDelegateImpl.shared
}

/// 测量类型
@objc
public enum IDOMeasureType: Int {
    /// 血压
    case bloodPressure = 0
    /// 心率
    case heartRate = 1
    /// 血氧
    case spo2 = 2
    /// 压力
    case stress = 3
    /// 一键测量
    case oneClick = 4
    /// 体温
    case temperature = 5
    /// 身体成分
    case bodyComposition = 6
}

/// 测量状态
@objc
public enum IDOMeasureStatus: Int {
    /// 不支持
    case notSupport = 0
    /// 正在测量
    case measuring = 1
    /// 测量成功
    case success = 2
    /// 测量失败
    case failed = 3
    /// 设备正在运动模式
    case deviceInSportMode = 4
    /// 返回测量中数据
    case measuringData = 5
    /// 超时 (APP)
    case timeout = 6
    /// 停止 (APP)
    case stopped = 7
    /// 断连 (APP)
    case disconnected = 8
    /// 未佩戴 (设备通知 72)
    case unworn = 9
}

/// 测量结果模型
@objcMembers
public class IDOMeasureResult: NSObject {
    /// 状态
    public var status: IDOMeasureStatus = .notSupport
    /// 收缩压 (高压)
    public var systolicBp: Int = 0
    /// 舒张压 (低压)
    public var diastolicBp: Int = 0
    /// 值 (心率/血氧/压力)
    public var value: Int = 0
    /// 一键测量-心率
    public var oneClickHr: Int = 0
    /// 一键测量-压力
    public var oneClickStress: Int = 0
    /// 一键测量-血氧
    public var oneClickSpo2: Int = 0
    /// 体温值 (单位摄氏度,该值扩大了10倍, 如 366 表示 36.6度)
    public var temperatureValue: Int = 0
    
    fileprivate static func fromPigeon(_ result: MeasureResult) -> IDOMeasureResult {
        let rs = IDOMeasureResult()
        rs.status = IDOMeasureStatus(rawValue: result.status?.rawValue ?? 0) ?? .notSupport
        rs.systolicBp = Int(result.systolicBp ?? 0)
        rs.diastolicBp = Int(result.diastolicBp ?? 0)
        rs.value = Int(result.value ?? 0)
        rs.oneClickHr = Int(result.oneClickHr ?? 0)
        rs.oneClickStress = Int(result.oneClickStress ?? 0)
        rs.oneClickSpo2 = Int(result.oneClickSpo2 ?? 0)
        rs.temperatureValue = Int(result.temperatureValue ?? 0)
        return rs
    }
}


@objcMembers
public class IDOMeasureManager {
    public static let shared = IDOMeasureManager()
    private init() {}
    
    /// 监听测量中数据（心率/血氧有效）
    public func listenProcessMeasureData(callback: @escaping (IDOMeasureResult) -> Void) {
        _measureImpl.callbackProcessMeasureData = { result in
            let rs = IDOMeasureResult.fromPigeon(result)
            callback(rs)
        }
    }
    
    /// 开始测量
    public func startMeasure(type: IDOMeasureType, completion: @escaping (Bool) -> Void) {
        let pType = MeasureType(rawValue: type.rawValue)!
        _measureMgr?.startMeasure(type: pType, completion: completion)
    }
    
    /// 停止测量
    public func stopMeasure(type: IDOMeasureType, completion: @escaping (Bool) -> Void) {
        let pType = MeasureType(rawValue: type.rawValue)!
        _measureMgr?.stopMeasure(type: pType, completion: completion)
    }
    
    /// 获取测量数据
    public func getMeasureData(type: IDOMeasureType, completion: @escaping (IDOMeasureResult) -> Void) {
        let pType = MeasureType(rawValue: type.rawValue)!
        _measureMgr?.getMeasureData(type: pType, completion: { result in
            completion(IDOMeasureResult.fromPigeon(result))
        })
    }
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/pigeon_implement/IDOMeasureManager.kt
package com.idosmart.pigeon_implement

import com.idosmart.pigeongen.api_measure.Measure
import com.idosmart.pigeongen.api_measure.MeasureResult
import com.idosmart.pigeongen.api_measure.MeasureType
import com.idosmart.pigeongen.api_measure.MeasureStatus
import com.idosmart.protocol_channel.ProtocolChannelPlugin
import com.idosmart.protocol_channel.innerRunOnMainThread

private val _measureMgr: Measure?
    get() = ProtocolChannelPlugin.instance().measure()

private val _measureImpl: MeasureDelegateImpl
    get() = MeasureDelegateImpl.instance()

open class IDOMeasureManager private constructor() {
    companion object {
        val shared: IDOMeasureManager by lazy { IDOMeasureManager() }
    }

    /**
     * 测量结果监听
     */
    fun listenProcessMeasureData(callback: (IDOMeasureResult) -> Unit) {
        _measureImpl.callbackProcessMeasureData = { result ->
            callback(IDOMeasureResult.fromPigeon(result))
        }
    }

    fun startMeasure(type: IDOMeasureType, completion: (Boolean) -> Unit) {
        innerRunOnMainThread {
            val pType = MeasureType.values().firstOrNull { it.raw == type.raw } ?: MeasureType.BLOODPRESSURE
            _measureMgr?.startMeasure(pType) {
                completion(it)
            }
        }
    }

    fun stopMeasure(type: IDOMeasureType, completion: (Boolean) -> Unit) {
        innerRunOnMainThread {
            val pType = MeasureType.values().firstOrNull { it.raw == type.raw } ?: MeasureType.BLOODPRESSURE
            _measureMgr?.stopMeasure(pType) {
                completion(it)
            }
        }
    }

    fun getMeasureData(type: IDOMeasureType, completion: (IDOMeasureResult) -> Unit) {
        innerRunOnMainThread {
            val pType = MeasureType.values().firstOrNull { it.raw == type.raw } ?: MeasureType.BLOODPRESSURE
            _measureMgr?.getMeasureData(pType) {
                completion(IDOMeasureResult.fromPigeon(it))
            }
        }
    }
}


open class IDOMeasureResult(
    var status: IDOMeasureStatus = IDOMeasureStatus.NOT_SUPPORT,
    var systolicBp: Int = 0,
    var diastolicBp: Int = 0,
    var value: Int = 0,
    var oneClickHr: Int = 0,
    var oneClickStress: Int = 0,
    var oneClickSpo2: Int = 0,
    var temperatureValue: Int = 0
) {
    companion object {
        internal fun fromPigeon(result: MeasureResult): IDOMeasureResult {
            return IDOMeasureResult(
                status = IDOMeasureStatus.values().firstOrNull { it.raw == (result.status?.raw ?: 0) } ?: IDOMeasureStatus.NOT_SUPPORT,
                systolicBp = result.systolicBp?.toInt() ?: 0,
                diastolicBp = result.diastolicBp?.toInt() ?: 0,
                value = result.value?.toInt() ?: 0,
                oneClickHr = result.oneClickHr?.toInt() ?: 0,
                oneClickStress = result.oneClickStress?.toInt() ?: 0,
                oneClickSpo2 = result.oneClickSpo2?.toInt() ?: 0,
                temperatureValue = result.temperatureValue?.toInt() ?: 0
            )
        }
    }
}

enum class IDOMeasureType(val raw: Int) {
    BLOOD_PRESSURE(0),
    HEART_RATE(1),
    SPO2(2),
    STRESS(3),
    ONE_CLICK(4),
    TEMPERATURE(5),
    BODY_COMPOSITION(6)
}

enum class IDOMeasureStatus(val raw: Int) {
    NOT_SUPPORT(0),
    MEASURING(1),
    SUCCESS(2),
    FAILED(3),
    DEVICE_IN_SPORT_MODE(4),
    MEASURING_DATA(5),
    TIMEOUT(6),
    STOPPED(7),
    DISCONNECTED(8),
    UNWORN(9)
}
